File: src/test/scala/edu/colorado/plv/fixr/tests/TestClassBase.scala
package edu.colorado.plv.fixr.tests

import soot.SootClass
import edu.colorado.plv.fixr.SootHelper
import soot.Scene
import org.scalatest.FunSuite
import org.scalatest.BeforeAndAfter

/**
 * Basic class to be used for testing.
 * 
 * @author: Sergio Mover
 * @mail: sergio.mover@colorado.edu
 */
class TestClassBase(classPath : String, testClassName : String,
    resClassName : String) extends FunSuite with BeforeAndAfter { 	
 	var testClass : SootClass = null;
 	var resClass : SootClass = null;
 	
 	def getTestClass() : SootClass = testClass;
 	def setTestClass(testClass : SootClass) : Unit = {
 	  this.testClass = testClass;
 	}

	def getResClass() : SootClass = resClass;
 	def setResClass(resClass : SootClass) : Unit = {
 	  this.resClass = resClass;
 	}

 	def getClassPath() : String = classPath;
 	
 	/**
 	 * Setup the soot environment loading the test class and the result class 
 	 *
 	 */
 	before {
 	  SootHelper.reset();			
		SootHelper.configure(getClassPath(), true);
		
		/* Load dependencies */
		if (null != testClassName) {
			Scene.v().addBasicClass(testClassName, SootClass.HIERARCHY);
		}
		if (null != resClassName) {
		  Scene.v().addBasicClass(resClassName, SootClass.HIERARCHY);
		}
		Scene.v().loadNecessaryClasses();		
		
		if (null != testClassName) {	  
		  testClass = Scene.v().tryLoadClass(testClassName, SootClass.BODIES);
		  assert(testClass != null)
		}
		
		if (null != resClassName) {	
		  resClass = Scene.v().tryLoadClass(resClassName, SootClass.BODIES);
		  assert(resClass != null)
		}
 	} 	 	
}

File: src/test/java/edu/colorado/plv/fixr/TestClassBase.java
package edu.colorado.plv.fixr;

import org.junit.Before;

import soot.Scene;
import soot.SootClass;

public abstract class TestClassBase {
	protected final String classPath = "src/test/resources/jimple";

	protected SootClass testClass = null;
	protected SootClass resClass = null;

	public TestClassBase() {	
	}

	/**
	 * @return the testClass
	 */
	public SootClass getTestClass() {
		return testClass;
	}

	/**
	 * @param testClass the testClass to set
	 */
	public void setTestClass(SootClass testClass) {
		this.testClass = testClass;
	}

	/**
	 * @return the resClass
	 */
	public SootClass getResClass() {
		return resClass;
	}

	/**
	 * @param resClass the resClass to set
	 */
	public void setResClass(SootClass resClass) {
		this.resClass = resClass;
	}
	
	public abstract String getTestClassName();	
	public abstract String getResClassName();	
	
	public String getClassPath() {
		return this.classPath;		
	}	
	
	@Before
	public void setup()
	{
		SootHelper.reset();			
		SootHelper.configure(getClassPath(), true);
		
		/* Load dependencies */
		if (null != getTestClassName()) {
			Scene.v().addBasicClass(getTestClassName(), SootClass.HIERARCHY);
		}
		if (null != getResClassName()) {
			Scene.v().addBasicClass(getResClassName(), SootClass.HIERARCHY);			
		}

		Scene.v().loadNecessaryClasses();		
		
		if (null != getTestClassName())
			setTestClass(Scene.v().tryLoadClass(getTestClassName(), SootClass.BODIES));			
		if (null != getResClassName())				
			setResClass(Scene.v().tryLoadClass(getResClassName(), SootClass.BODIES));
	}	
}
